votingpool/db.go

```diff
@@ -56,13 +59,79 @@ type dbSeriesRow struct {
 	privKeysEncrypted [][]byte
 }
 
+// getUsedAddrBucketID returns the used addresses bucket ID for the given series
+// and branch. It has the form seriesID:branch.
+func getUsedAddrBucketID(seriesID uint32, branch Branch) []byte {
+	return bytes.Join(
+		[][]byte{uint32ToBytes(seriesID), uint32ToBytes(uint32(branch))}, []byte(":"))
+}
+
+// putUsedAddrHash adds an entry (key==index, value==encryptedHash) to the used
+// addresses bucket of the given pool, series and branch.
+func putUsedAddrHash(tx walletdb.Tx, poolID []byte, seriesID uint32, branch Branch,
+	index Index, encryptedHash []byte) error {
+
+	usedAddrs := tx.RootBucket().Bucket(poolID).Bucket(usedAddrsBucketName)
+	bucket, err := usedAddrs.CreateBucketIfNotExists(getUsedAddrBucketID(seriesID, branch))
+	if err != nil {
+		return newError(ErrDatabase, "failed to store used address hash", err)
+	}
+	return bucket.Put(uint32ToBytes(uint32(index)), encryptedHash)
+}
+
+// getUsedAddrHash returns the addr hash with the given index from the used
+// addresses bucket of the given pool, series and branch.
+func getUsedAddrHash(
+	tx walletdb.Tx, poolID []byte, seriesID uint32, branch Branch, index Index) []byte {
+
+	usedAddrs := tx.RootBucket().Bucket(poolID).Bucket(usedAddrsBucketName)
+	bucket := usedAddrs.Bucket(getUsedAddrBucketID(seriesID, branch))
+	if bucket == nil {
+		return nil
+	}
+	return bucket.Get(uint32ToBytes(uint32(index)))
+}
+
+// getMaxUsedIdx returns the highest used index from the used addresses bucket
+// of the given pool, series and branch.
+func getMaxUsedIdx(tx walletdb.Tx, poolID []byte, seriesID uint32, branch Branch) (Index, error) {
+	maxIdx := Index(0)
+	usedAddrs := tx.RootBucket().Bucket(poolID).Bucket(usedAddrsBucketName)
+	bucket := usedAddrs.Bucket(getUsedAddrBucketID(seriesID, branch))
+	if bucket == nil {
+		return maxIdx, nil
+	}
+	err := bucket.ForEach(
```

Comment:
This will have severe scalability issues searching all of the keys each time.  Imagine when you start to have a few million addresses.

What do you think about keeping a key in a metadata bucket which is updated as necessary? I think that would be the most efficient.

Alternatively, if you don't want to keep a field updated, perhaps a large gap linear forward search + binary backwards search?  For example, check for 1000000, 2000000, 3000000, .... until it doesn't exist, and then use a binary search to find the max.  With this type of scheme, having say 50 million addresses would result in a worst case scenario of testing for 50 + log_2(50,000,000) = 70 keys instead of 50 million.